### PR TITLE
Device icon updates to the appropriate icon when changed

### PR DESCRIPTION
**Pull request by JohnVUgonabo**
## Purpose

* Fixed the device icon to update to the appropriate icon when changed, in both the lobby and calling views

## Does this introduce a breaking change?

```
[ ] Yes
[ x ] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[ x ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What to Check
Verify that the following are valid
* Changing the device in the both the lobby and calling view reflects appropriate change to device icon
* When joining call, device selection from lobby is reflected in calling view

## Other Information

Link to bug https://skype.visualstudio.com/SPOOL/_workitems/edit/2508580

### CODE REVIEW

File: AzureCalling/Controllers/CallViewController.swift [hunk 1]
```diff
@@ -629,6 +631,21 @@ class CallViewController: UIViewController, UICollectionViewDelegate, UICollecti
             bottomDrawerViewController.refreshBottomDrawer()
         }
     }
+    // update the audio device icon with current selection
+    private func updateAudioDevice() {
```
Comment (jimchou-dev) [inline AzureCalling/Controllers/CallViewController.swift:635]: naming too generic it's not actually updating the audio device, should probably include update button icon in the name

File: AzureCalling/Controllers/LobbyViewController.swift [hunk 2]
```diff
@@ -243,6 +245,19 @@ class LobbyViewController: UIViewController, UITextFieldDelegate {
...
+    private func updateAudioDevice() {
+        let currentAudioDeviceType = AudioSessionManager.getCurrentAudioDeviceType() // Get the current audio selection
+        // set icon based on audio type
+        var deviceIcon: UIImage
+        switch currentAudioDeviceType {
+        case .receiver:
+            deviceIcon = UIImage(named: "ic_fluent_speaker_2_28_regular")!
+        case .speaker:
+            deviceIcon = UIImage(named: "ic_fluent_speaker_2_28_filled")!
+        }
```
Comment (jimchou-dev) [inline AzureCalling/Controllers/LobbyViewController.swift:258]: Similar code with different naming for `deviceIcon` and `image` with CallViewController. updateAudioDevice

Can extract a function to return the UIImage

File: AzureCalling/Controllers/CallViewController.swift [hunk 3]
```diff
@@ -64,6 +64,7 @@ class CallViewController: UIViewController, UICollectionViewDelegate, UICollecti
         toggleMuteButton.isSelected = joinCallConfig.isMicrophoneMuted
 
         updateToggleVideoButtonState()
+        updateAudioDevice() // update audio icon at start of call
```
Comment (jimchou-dev) [inline AzureCalling/Controllers/CallViewController.swift:67]: Remove the comments (applies for all the files)

File: AzureCalling/Controllers/datasource/AudioDeviceSelectionDataSource.swift [hunk 4]
```diff
@@ -71,3 +72,8 @@ class AudioDeviceSelectionDataSource: NSObject, BottomDrawerDataSource {
         return cell
     }
 }
+
+//this handles updating the device icon when changed
+protocol AudioDeviceSelectionViewControllerDelegate: AnyObject {
+    func updateDeviceAudioSelection() // updates the device audio icon with current Selection
```
Comment (jimchou-dev) [inline AzureCalling/Controllers/datasource/AudioDeviceSelectionDataSource.swift:78]: Why is this `DeviceAudio` when other places are `audioDevice`. Maybe rename to updateAudioDeviceSelectionButtonIcon or something?

**Comment by JohnVUgonabo** [comment]
@jimchou-dev Thanks for the pointers and corrections, I would make the necessary changes and update this PR

File: AzureCalling/Controllers/LobbyViewController.swift [hunk 5]
```diff
@@ -124,6 +125,7 @@ class LobbyViewController: UIViewController, UITextFieldDelegate {
 
     private func openAudioDeviceDrawer() {
         let audioDeviceSelectionDataSource = AudioDeviceSelectionDataSource()
+        audioDeviceSelectionDataSource.audioDeviceSelectionDelegate = self
```
Comment (AmyL219) [inline AzureCalling/Controllers/LobbyViewController.swift:128]: I think it's more consistent with the original pattern by using a closure here. 
similar to `dismissDrawer` inside `AudioDeviceSelectionDataSource`, we could have a closure which is defined as 
`var didSelectedItem: (BottomDrawerItem) -> void`  using initializer to define the closure
Inside this closure,  we update the icon on main thread
```
audioDeviceSelectionDataSource.didSelectedItem = {[weak self] item
guard let self = self else {return }
updateAudioDeviceButtonIcon(item.avatar)
}
```

Comment (JohnVUgonabo) [reply AzureCalling/Controllers/LobbyViewController.swift]: @AmyL219 Thank you for the feedback and suggestions. Using a closure was a much better approach, trimmed down the amount of codes written. PR now updated

**Review by jimchou-dev — DISMISSED** [review]
Tested, LGTM

File: AzureCalling/Controllers/CallViewController.swift [hunk 6]
```diff
@@ -151,6 +152,9 @@ class CallViewController: UIViewController, UICollectionViewDelegate, UICollecti
 
     private func openAudioDeviceDrawer() {
         let audioDeviceSelectionDataSource = AudioDeviceSelectionDataSource()
+        audioDeviceSelectionDataSource.didSelectAudioDevice = {
+            self.updateAudioDeviceButtonIcon()
```
Comment (AmyL219) [inline AzureCalling/Controllers/CallViewController.swift:156]: Let's use [weak self] here  which prevent retain cycle
audioDeviceSelectionDataSource.didSelectAudioDevice = { [weak self] in 
guard let self = self else {return}
            self.updateAudioDeviceButtonIcon()

File: AzureCalling/Controllers/LobbyViewController.swift [hunk 7]
```diff
@@ -243,6 +247,11 @@ class LobbyViewController: UIViewController, UITextFieldDelegate {
         startCallButton.isEnabled = isDisplayNameValid && isPermissionValid
     }
 
+    private func updateAudioDeviceButtonIcon() {
+        let deviceIcon = AudioSessionManager.selectedAudioDeviceButtonIcon()
```
Comment (AmyL219) [inline AzureCalling/Controllers/LobbyViewController.swift:251]: same weak self here

File: AzureCalling/Utilities/AudioSessionManager.swift [hunk 8]
```diff
@@ -46,4 +46,15 @@ class AudioSessionManager {
             }
         }
     }
+    public static func selectedAudioDeviceButtonIcon() -> UIImage {
```
Comment (AmyL219) [inline AzureCalling/Utilities/AudioSessionManager.swift:49]: Not sure if it's a good idea to read the icon image from AudioSessionManager. 
Maybe extend the `AudioDeviceType` to be an enum with associated value of iconImage. Rather to put this function here.

Comment (JohnVUgonabo) [reply AzureCalling/Utilities/AudioSessionManager.swift]: @AmyL219 Thanks for the suggestion. It's a good idea which also helps AudioSessionManager maintain its core and cleans up the code. Suggestions now implemented and PR updated.

**Review by AmyL219 — APPROVED** [review]

**Review by jimchou-dev — APPROVED** [review]

**Pull request merged**